src/main/groovy/com/blackducksoftware/integration/hub/detect/DetectProjectManager.groovy

```diff
@@ -193,27 +197,39 @@ class DetectProjectManager implements SummaryResultReporter {
         return detectSummaryResults;
     }
 
-    private String createBdioFilename(BomToolType bomToolType, String finalSourcePathPiece, String projectName, String projectVersionName) {
-        def names = [finalSourcePathPiece, projectName, projectVersionName]
-        names.sort { -it.size() }
-        String filename = generateFilename(bomToolType, finalSourcePathPiece, projectName, projectVersionName)
-        for (int i = 0; (filename.length() >= 255) && (i < 3); i++) {
-            names[i] = DigestUtils.sha1Hex(names[i])
-            if (names[i].length() > 15) {
-                names[i] = names[i].substring(0, 15)
-            }
+    private String generateShortenedFilename(BomToolType bomToolType, String finalSourcePathPiece, ExternalId externalId) {
+
+        List<String> filenamePieces = new ArrayList<>();
+        filenamePieces.addAll(externalId.getExternalIdPieces());
+        filenamePieces.add(finalSourcePathPiece);
+        String filename = generateFilename(bomToolType, filenamePieces)
 
-            filename = generateFilename(bomToolType, names[0], names[1], names[2])
+        if (filename.length() >= 255){
+            filenamePieces.sort { it.size() }
+            for (int i = filenamePieces.size() - 1; (filename.length() >= 255) && (i >= 0); i--) {
+                filenamePieces[i] = DigestUtils.sha1Hex(filenamePieces[i])
+                if (filenamePieces[i].length() > 15) {
+                    filenamePieces[i] = filenamePieces[i].substring(0, 15)
+                }
+
+                filename = generateFilename(bomToolType, filenamePieces)
+            }
         }
 
         filename
     }
 
-    private String generateFilename(BomToolType bomToolType, String finalSourcePathPiece, String projectName, String projectVersionName) {
-        List<String> safePieces = [bomToolType.toString(), projectName, projectVersionName, finalSourcePathPiece, 'bdio'].collect { integrationEscapeUtil.escapeForUri(it) }
+    private String generateFilename(BomToolType bomToolType, List<String> pieces) {
+        List<String> rawPieces = new ArrayList<String>();
+        rawPieces.add(bomToolType.toString());
+        rawPieces.addAll(pieces);
+        rawPieces.add('bdio');
 
-        String filename = (safePieces as Iterable).join('_') + '.jsonld'
-        filename
+        List<String> safePieces = rawPieces.collect { integrationEscapeUtil.escapeForUri(it) };
+
+        String filename = (safePieces as Iterable).join('_') + '.jsonld';
+
```

Comment:
I know Iterable was here before...do we know why? It seems odd to me - I thought join was available to a List. Am I off base here?

Reply:
When I try to remove that it says:
Deprecated.  Use the Iterable version of join instead